File: tools-python/bridge_report_tools/contracts/annual_inspection.py
"""模块 03 的年度检测候选数据契约。

这里的模型是 Word 解析、C++ 保存候选 JSON、前端校对和后续确认入库之间的共同边界。
它描述的是“候选数据”，不是已经写入 PostgreSQL 的正式事实。
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


ReviewStatus = Literal["待确认", "已确认", "已修改", "已忽略"]
DefectGroupReviewStatus = Literal["待确认", "已确认"]
ComparisonConfirmationStatus = Literal["待确认", "已确认", "已修改", "已拒绝"]
Severity = Literal["info", "warning", "error"]
StructurePart = Literal["全桥", "上部结构", "下部结构", "桥面系", "其他"]
SourceType = Literal["软件导出Word", "正式Word", "Excel病害表", "图片包", "接口同步", "JSON导入"]
FileRole = Literal["当前年度检测资料", "历史正式报告", "历史基线资料", "修订资料"]
DataRole = Literal["当前年度", "历史基线", "修订版"]
BridgeMatchStatus = Literal["匹配", "不匹配", "待人工确认"]
ComparisonType = Literal[
    "原病害无明显变化",
    "原病害发展",
    "原病害减轻",
    "原病害修复",
    "新增病害",
    "原病害未见",
    "无法判断",
]

class ContractModel(BaseModel):
    """所有契约模型默认禁止额外字段，避免解析器悄悄输出未评审的数据。"""

    model_config = ConfigDict(extra="forbid")


class WarningItem(ContractModel):
    code: str
    message: str
    severity: Severity
    target_candidate_id: str | None = None


class SourceRef(ContractModel):
    """候选对象的来源证据，用于人工校对时回到 Word 表格行或段落。"""

    source_type: Literal["word", "manual"] = "word"
    chapter: str | None = None
    table_title: str | None = None
    table_index: int | None = Field(default=None, ge=0)
    row_index: int | None = Field(default=None, ge=0)
    column_name: str | None = None
    raw_row_text: str | None = None
    photo_area_caption: str | None = None
    file_role: FileRole | None = None
    paragraph_index: int | None = Field(default=None, ge=0)


class ContractInfo(ContractModel):
    name: Literal["BridgeAnnualInspectionData"]
    version: Literal["5.0"]
    generated_at: datetime
    producer: str
    parser_name: str
    parser_version: str


class ImportContext(ContractModel):
    source_type: SourceType
    file_role: FileRole
    archived_file_system_number: str
    import_record_system_number: str


class BridgeCheck(ContractModel):
    selected_bridge_system_number: str
    extracted_bridge_name: str | None = None
    match_status: BridgeMatchStatus
    warnings: list[WarningItem]


class InspectionInfo(ContractModel):
    inspection_year: int = Field(ge=1900, le=2200)
    inspection_date: date
    report_number: str
    project_name: str
    data_role: DataRole


class Measurement(ContractModel):
    dimension_type: str
    value_type: Literal["single", "range"]
    value: float | None = None
    minimum_value: float | None = None
    maximum_value: float | None = None
    unit: str
    is_approximate: bool = False
    source_text: str

    @model_validator(mode="after")
    def validate_value_shape(self) -> Measurement:
        if self.value_type == "single":
            if self.value is None:
                raise ValueError("single measurement requires value")
            if self.minimum_value is not None or self.maximum_value is not None:
                raise ValueError("single measurement must not contain range endpoints")
            return self

        if self.value is not None:
            raise ValueError("range measurement must not contain value")
        if self.minimum_value is None or self.maximum_value is None:
            raise ValueError("range measurement requires both endpoints")
        if self.minimum_value > self.maximum_value:
            raise ValueError("range measurement minimum must not exceed maximum")
        return self


class PhotoReference(ContractModel):
    """病害对一张照片的引用，以及人工核对结论。

    `photo_number` 只是 Word 导入路的匹配键：Word 病害表的「照片编号」列和图片题注
    靠这个字符串对上，除它之外没有别的线索。来源软件导入没有这个概念——照片用外键
    直接绑在病害上（`images.ForeignKey`），所以编号为空，配对一律走
    `photo_candidate_id`。

    唯一的例外是 `pending` 和 `missing`：这两种状态下照片实体还不存在，
    `photo_candidate_id` 必为空，编号是这条引用仅有的标识，不能省。

    放宽为可空没有提高契约版本：5.0 下合法的文档现在依然合法，只是新文档允许省略
    该字段。升版会让库里已存的 5.0 草稿在重新打开时被拒。
    """

    photo_number: str | None = None
    resolution: Literal["pending", "matched", "relinked", "missing", "unrelated"]
    photo_candidate_id: str | None = None
    resolved_defect_candidate_id: str | None = None
    review_note: str | None = None

    @model_validator(mode="after")
    def validate_resolution_targets(self) -> PhotoReference:
        has_photo = self.photo_candidate_id is not None
        has_defect = self.resolved_defect_candidate_id is not None
        valid_targets = {
            "pending": not has_photo and not has_defect,
            "matched": has_photo and has_defect,
            "relinked": has_photo and has_defect,
            "missing": not has_photo and not has_defect,
            "unrelated": has_photo and not has_defect,
        }
        if not valid_targets[self.resolution]:
            raise ValueError(
                f"photo reference targets are invalid for resolution {self.resolution}"
            )
        if self.photo_number is not None and not self.photo_number.strip():
            raise ValueError("photo_number must be omitted rather than left blank")
        if self.resolution in ("pending", "missing") and self.photo_number is None:
            raise ValueError(
                f"photo reference with resolution {self.resolution} has no photo to point at,"
                " so photo_number is required"
            )
        return self


class DefectCandidate(ContractModel):
    """第二章结构病害检查表中的一条病害候选记录。"""

    candidate_id: str
    source_structure_part: StructurePart | None = None
    component_name: str
    component_number: str | None = None
    defect_type: str
    defect_location: str
    defect_scale: int | None = Field(default=None, gt=0, strict=True)
    defect_description: str
    quantity_text: str | None = None
    measurement_text: str | None = None
    measurements: list[Measurement]
    #: 来源软件原始评定分组和指标身份。ID 来自 judgeTreeId / judgeIndexId，编号来自
    #: judgeTree.chapterNum / judgeIndex.tableNum。它们与派生的 H21 指标严格分开，重绑与
    #: 草稿保存都不得清空。
    source_defect_group_id: str | None = None
    source_defect_group_number: str | None = None
    source_defect_indicator_id: str | None = None
    source_defect_indicator_number: str | None = None
    photo_references: list[PhotoReference]
    group_review_status: DefectGroupReviewStatus
    severity: Severity | None = None
    remark: str | None = None
    source_ref: SourceRef
    confidence: float = Field(ge=0, le=1)
    review_status: ReviewStatus
    warnings: list[WarningItem]

    @field_validator("photo_references")
    @classmethod
    def require_unique_photo_references(
        cls, value: list[PhotoReference]
    ) -> list[PhotoReference]:
        # 只比对有编号的引用：来源软件那条路整条链都没有编号，若把 None 也算进去，
        # 一条病害挂两张图就会被误判成重复。
        photo_numbers = [
            reference.photo_number for reference in value if reference.photo_number is not None
        ]
        if len(photo_numbers) != len(set(photo_numbers)):
            raise ValueError("photo_references.photo_number must be unique")
        # 没有编号时，photo_candidate_id 才是这条引用的身份，同样不许重复。
        candidate_ids = [
            reference.photo_candidate_id
            for reference in value
            if reference.photo_candidate_id is not None
        ]
        if len(candidate_ids) != len(set(candidate_ids)):
            raise ValueError("photo_references.photo_candidate_id must be unique")
        return value

    @field_validator(
        "source_defect_group_id",
        "source_defect_group_number",
        "source_defect_indicator_id",
        "source_defect_indicator_number",
    )
    @classmethod
    def require_non_empty_source_reference(cls, value: str | None) -> str | None:
        if value is not None and not value.strip():
            raise ValueError("source indicator references must not be blank")
        return value


class ExtractedPhotoFile(ContractModel):
    temporary_file_name: str
    original_caption: str | None = None
    archive_relative_path: str | None = None


class PhotoCandidate(ContractModel):
    candidate_id: str
    #: Word 路是图片题注里的编号；来源软件路没有编号，为空。见 PhotoReference 的说明。
    photo_number: str | None = None
    linked_defect_candidate_id: str | None = None
    extracted_file: ExtractedPhotoFile
    source_ref: SourceRef
    confidence: float = Field(ge=0, le=1)
    warnings: list[WarningItem]


class ComparisonMatchBasis(ContractModel):
    same_component: bool
    same_defect_type: bool
    location_similarity: float = Field(ge=0, le=1)
    measurement_change_detected: bool
    photo_number_related: bool


class ComparisonCandidate(ContractModel):
    """事实入库后生成的历史对比候选，不由 Word 解析器直接产生。"""

    candidate_id: str
    previous_defect_observation_system_number: str | None = None
    current_defect_observation_system_number: str | None = None
    comparison_type: ComparisonType
    match_basis: ComparisonMatchBasis | None = None
    change_summary: str | None = None
    confidence: float = Field(ge=0, le=1)
    confirmation_status: ComparisonConfirmationStatus
    review_note: str | None = None
    warnings: list[WarningItem]


class ReportTextCandidate(ContractModel):
    """正式报告文本抽取的预留扩展口，第一版不作为病害事实来源。"""

    candidate_id: str
    section_key: str
    section_title: str
    text: str
    usage: str | None = None
    source_ref: SourceRef | None = None
    review_status: ReviewStatus


class BridgeAnnualInspectionData(ContractModel):
    """一次导入任务的完整候选 JSON。"""

    contract: ContractInfo
    import_context: ImportContext
    bridge_check: BridgeCheck
    inspection: InspectionInfo
    defects: list[DefectCandidate]
    photos: list[PhotoCandidate]
    comparison_candidates: list[ComparisonCandidate]
    report_text_candidates: list[ReportTextCandidate]
    warnings: list[WarningItem]
    errors: list[WarningItem]

File: frontend/src/contracts/annualInspection.ts
export type ReviewStatus = "待确认" | "已确认" | "已修改" | "已忽略";
export type DefectGroupReviewStatus = "待确认" | "已确认";
export type ComparisonConfirmationStatus = "待确认" | "已确认" | "已修改" | "已拒绝";
export type Severity = "info" | "warning" | "error";
export type StructurePart = "全桥" | "上部结构" | "下部结构" | "桥面系" | "其他";
export type SourceType = "软件导出Word" | "正式Word" | "Excel病害表" | "图片包" | "接口同步" | "JSON导入";
export type FileRole = "当前年度检测资料" | "历史正式报告" | "历史基线资料" | "修订资料";
export type DataRole = "当前年度" | "历史基线" | "修订版";
export type BridgeMatchStatus = "匹配" | "不匹配" | "待人工确认";
export type PhotoMatchStatus = "高置信候选" | "待校对" | "已确认" | "未关联" | "已忽略";
export type ComparisonType =
  | "原病害无明显变化"
  | "原病害发展"
  | "原病害减轻"
  | "原病害修复"
  | "新增病害"
  | "原病害未见"
  | "无法判断";

// 以下接口与 Python Pydantic 模型保持同名字段，作为前端校对页面的候选数据边界。
export interface WarningItem {
  code: string;
  message: string;
  severity: Severity;
  target_candidate_id?: string | null;
}

export interface SourceRef {
  source_type?: "word" | "manual";
  chapter?: string | null;
  table_title?: string | null;
  table_index?: number | null;
  row_index?: number | null;
  column_name?: string | null;
  raw_row_text?: string | null;
  photo_area_caption?: string | null;
  file_role?: FileRole | null;
  paragraph_index?: number | null;
}

export interface ContractInfo {
  name: "BridgeAnnualInspectionData";
  version: "3.0";
  generated_at: string;
  producer: string;
  parser_name: string;
  parser_version: string;
}

export interface ImportContext {
  source_type: SourceType;
  file_role: FileRole;
  archived_file_system_number: string;
  import_record_system_number: string;
}

export interface BridgeCheck {
  selected_bridge_system_number: string;
  extracted_bridge_name?: string | null;
  match_status: BridgeMatchStatus;
  warnings: WarningItem[];
}

export interface InspectionInfo {
  inspection_year: number;
  inspection_date: string;
  report_number: string;
  project_name: string;
  data_role: DataRole;
}

export interface Measurement {
  dimension_type: string;
  value_type: "single" | "range";
  value: number | null;
  minimum_value: number | null;
  maximum_value: number | null;
  unit: string;
  is_approximate: boolean;
  source_text: string;
}

export interface RangeSplitOrigin {
  operation_id: string;
  source_candidate_id: string;
  source_component_number: string;
  expanded_component_number: string;
  split_index: number;
  split_count: number;
  operated_by_user_id: string;
  operated_at: string;
}

export type PhotoReferenceResolution = "pending" | "matched" | "relinked" | "missing" | "unrelated";

/**
 * 评定树病害的匹配方式。前三种是后端分层确定性匹配写入的自动结果，`fuzzy_candidate`
 * 只是候选提示（永远不落 rating_tree_node_id），`manual` 是人工选择且不被自动结果覆盖。
 */
export const RATING_TREE_MATCH_METHODS = [
  "exact",
  "controlled_alias",
  "controlled_keyword",
  "fuzzy_candidate",
  "manual",
] as const;

export type RatingTreeMatchMethod = (typeof RATING_TREE_MATCH_METHODS)[number];

export interface PhotoReference {
  photo_number: string;
  resolution: PhotoReferenceResolution;
  photo_candidate_id: string | null;
  resolved_defect_candidate_id: string | null;
  review_note: string | null;
}

export interface DefectCandidate {
  candidate_id: string;
  source_structure_part?: StructurePart | null;
  component_name: string;
  component_number?: string | null;
  bridge_component_id?: string | null;
  standard_component_category_id?: string | null;
  resolved_structure_part?: StructurePart | null;
  component_inventory_revision_id?: string | null;
  component_match_candidate_ids?: string[];
  component_match_method?: "exact" | "confirmed_alias" | "normalized_candidate" | "manual" | "missing" | null;
  component_match_confirmed_by?: string | null;
  defect_type: string;
  defect_location: string;
  defect_scale?: number | null;
  defect_description: string;
  quantity_text?: string | null;
  measurement_text?: string | null;
  measurements: Measurement[];
  rating_tree_version_id?: string | null;
  rating_tree_node_id?: string | null;
  rating_tree_match_method?: RatingTreeMatchMethod | null;
  rating_tree_match_evidence?: string | null;
  standard_defect_indicator_id?: string | null;
  photo_references: PhotoReference[];
  group_review_status: DefectGroupReviewStatus;
  severity?: Severity | null;
  remark?: string | null;
  source_ref: SourceRef;
  confidence: number;
  review_status: ReviewStatus;
  review_note?: string | null;
  range_split_origin?: RangeSplitOrigin | null;
  warnings: WarningItem[];
}

export interface ExtractedPhotoFile {
  temporary_file_name: string;
  original_caption?: string | null;
  archive_relative_path?: string | null;
}

export interface PhotoCandidate {
  candidate_id: string;
  photo_number: string;
  linked_defect_candidate_id?: string | null;
  extracted_file: ExtractedPhotoFile;
  match_status: PhotoMatchStatus;
  source_ref: SourceRef;
  confidence: number;
  review_status: ReviewStatus;
  warnings: WarningItem[];
}

// 对比候选来自已确认事实之间的匹配，不是 Word 解析器直接抽取的内容。
export interface ComparisonMatchBasis {
  same_component: boolean;
  same_defect_type: boolean;
  location_similarity: number;
  measurement_change_detected: boolean;
  photo_number_related: boolean;
}

export interface ComparisonCandidate {
  candidate_id: string;
  previous_defect_observation_system_number?: string | null;
  current_defect_observation_system_number?: string | null;
  comparison_type: ComparisonType;
  match_basis?: ComparisonMatchBasis | null;
  change_summary?: string | null;
  confidence: number;
  confirmation_status: ComparisonConfirmationStatus;
  review_note?: string | null;
  warnings: WarningItem[];
}

export interface ReportTextCandidate {
  candidate_id: string;
  section_key: string;
  section_title: string;
  text: string;
  usage?: string | null;
  source_ref: SourceRef;
  review_status: ReviewStatus;
}

export interface BridgeAnnualInspectionData {
  contract: ContractInfo;
  import_context: ImportContext;
  bridge_check: BridgeCheck;
  inspection: InspectionInfo;
  defects: DefectCandidate[];
  photos: PhotoCandidate[];
  comparison_candidates: ComparisonCandidate[];
  report_text_candidates: ReportTextCandidate[];
  warnings: WarningItem[];
  errors: WarningItem[];
}

// 运行时校验只做前端入口防线，完整契约仍以 JSON Schema 和后端校验为准。
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hasOwn(value: Record<string, unknown>, property: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, property);
}

// 置信度用于校对排序和风险提示，必须是闭区间 0 到 1 的数值。
function isNumberFromZeroToOne(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1;
}

// 必填数组必须显式存在，空数组表示“确认没有”，缺字段表示契约不完整。
function getRequiredArray(value: Record<string, unknown>, property: string): unknown[] | null {
  const member = value[property];
  return hasOwn(value, property) && Array.isArray(member) ? member : null;
}

// 必填对象必须显式存在，保证前端拿到的是稳定的候选 JSON 骨架。
function getRequiredObject(value: Record<string, unknown>, property: string): Record<string, unknown> | null {
  const member = value[property];
  return hasOwn(value, property) && isRecord(member) ? member : null;
}

function hasValidConfidence(value: unknown): boolean {
  return isRecord(value) && isNumberFromZeroToOne(value.confidence);
}

function hasRequiredArrayMembers(value: Record<string, unknown>, members: string[]): boolean {
  return members.every((member) => getRequiredArray(value, member) !== null);
}

function hasRequiredObjectMembers(value: Record<string, unknown>, members: string[]): boolean {
  return members.every((member) => getRequiredObject(value, member) !== null);
}

function isValidSourceRef(value: unknown): boolean {
  return (
    isRecord(value) &&
    (value.source_type === undefined || value.source_type === "word" || value.source_type === "manual")
  );
}

// 病害标度是规范意义的正整数，与提示级别 severity 完全分离。
function isNullablePositiveInteger(value: unknown): boolean {
  return (
    value === undefined ||
    value === null ||
    (typeof value === "number" && Number.isInteger(value) && value > 0)
  );
}

function isValidMeasurement(value: unknown): value is Measurement {
  if (
    !isRecord(value) ||
    typeof value.dimension_type !== "string" ||
    value.dimension_type.length === 0 ||
    typeof value.unit !== "string" ||
    value.unit.length === 0 ||
    typeof value.source_text !== "string" ||
    value.source_text.length === 0 ||
    typeof value.is_approximate !== "boolean"
  ) {
    return false;
  }
  if (value.value_type === "single") {
    return typeof value.value === "number" && Number.isFinite(value.value) &&
      value.minimum_value === null && value.maximum_value === null;
  }
  if (value.value_type === "range") {
    return value.value === null &&
      typeof value.minimum_value === "number" && Number.isFinite(value.minimum_value) &&
      typeof value.maximum_value === "number" && Number.isFinite(value.maximum_value) &&
      value.minimum_value <= value.maximum_value;
  }
  return false;
}

function isValidRangeSplitOrigin(value: unknown): value is RangeSplitOrigin {
  if (!isRecord(value)) return false;
  const allowed = new Set([
    "operation_id",
    "source_candidate_id",
    "source_component_number",
    "expanded_component_number",
    "split_index",
    "split_count",
    "operated_by_user_id",
    "operated_at",
  ]);
  return (
    Object.keys(value).every((key) => allowed.has(key)) &&
    typeof value.operation_id === "string" && value.operation_id.length > 0 &&
    typeof value.source_candidate_id === "string" && value.source_candidate_id.length > 0 &&
    typeof value.source_component_number === "string" && value.source_component_number.length > 0 &&
    typeof value.expanded_component_number === "string" && value.expanded_component_number.length > 0 &&
    Number.isInteger(value.split_index) && (value.split_index as number) >= 1 &&
    Number.isInteger(value.split_count) && (value.split_count as number) >= 2 &&
    (value.split_index as number) <= (value.split_count as number) &&
    typeof value.operated_by_user_id === "string" && value.operated_by_user_id.length > 0 &&
    typeof value.operated_at === "string" && value.operated_at.length > 0
  );
}

function isValidPhotoReference(value: unknown): value is PhotoReference {
  if (
    !isRecord(value) ||
    typeof value.photo_number !== "string" ||
    value.photo_number.length === 0 ||
    !["pending", "matched", "relinked", "missing", "unrelated"].includes(String(value.resolution))
  ) {
    return false;
  }
  const hasPhoto = typeof value.photo_candidate_id === "string";
  const hasDefect = typeof value.resolved_defect_candidate_id === "string";
  const nullableIdsAreValid =
    (value.photo_candidate_id === null || hasPhoto) &&
    (value.resolved_defect_candidate_id === null || hasDefect);
  if (!nullableIdsAreValid) {
    return false;
  }
  if (value.resolution === "pending" || value.resolution === "missing") {
    return !hasPhoto && !hasDefect;
  }
  if (value.resolution === "matched" || value.resolution === "relinked") {
    return hasPhoto && hasDefect;
  }
  return hasPhoto && !hasDefect;
}

function isValidDefectCandidate(value: unknown): boolean {
  if (!isRecord(value) || !hasValidConfidence(value)) {
    return false;
  }
  const photoReferences = getRequiredArray(value, "photo_references");
  const measurements = getRequiredArray(value, "measurements");
  const matchCandidateIds = value.component_match_candidate_ids;
  const matchMethod = value.component_match_method;
  const rangeSplitOrigin = value.range_split_origin;
  return (
    hasRequiredArrayMembers(value, ["measurements", "photo_references", "warnings"]) &&
    measurements !== null && measurements.every(isValidMeasurement) &&
    hasRequiredObjectMembers(value, ["source_ref"]) &&
    isValidSourceRef(value.source_ref) &&
    typeof value.component_name === "string" &&
    value.component_name.length > 0 &&
    // 病害类型与位置允许为空串：Word 里的 "/"（本列无此项）被导入清洗成空业务值，
    // 受控关键词层再从描述兜底。描述本身仍然必须有内容。
    typeof value.defect_type === "string" &&
    typeof value.defect_location === "string" &&
    typeof value.defect_description === "string" &&
    value.defect_description.length > 0 &&
    !hasOwn(value, "structure_part") &&
    !hasOwn(value, "component_alias") &&
    !hasOwn(value, "defect_deduction") &&
    !hasOwn(value, "photo_numbers") &&
    !hasOwn(value, "confirmed_missing_photo_numbers") &&
    (value.group_review_status === "待确认" || value.group_review_status === "已确认") &&
    isNullablePositiveInteger(value.defect_scale) &&
    (matchCandidateIds === undefined ||
      (Array.isArray(matchCandidateIds) &&
        matchCandidateIds.every((item) => typeof item === "string") &&
        new Set(matchCandidateIds).size === matchCandidateIds.length)) &&
    (matchMethod === undefined ||
      matchMethod === null ||
      matchMethod === "exact" ||
      matchMethod === "confirmed_alias" ||
      matchMethod === "normalized_candidate" ||
      matchMethod === "manual" ||
      matchMethod === "missing") &&
    (rangeSplitOrigin === undefined ||
      rangeSplitOrigin === null ||
      isValidRangeSplitOrigin(rangeSplitOrigin)) &&
    (value.component_inventory_revision_id === undefined ||
      value.component_inventory_revision_id === null ||
      typeof value.component_inventory_revision_id === "string") &&
    (value.component_match_confirmed_by === undefined ||
      value.component_match_confirmed_by === null ||
      typeof value.component_match_confirmed_by === "string") &&
    (value.rating_tree_version_id === undefined ||
      value.rating_tree_version_id === null ||
      (typeof value.rating_tree_version_id === "string" &&
        value.rating_tree_version_id.trim().length > 0)) &&
    (value.rating_tree_node_id === undefined ||
      value.rating_tree_node_id === null ||
      (typeof value.rating_tree_node_id === "string" &&
        value.rating_tree_node_id.trim().length > 0)) &&
    (value.rating_tree_match_method === undefined ||
      value.rating_tree_match_method === null ||
      RATING_TREE_MATCH_METHODS.includes(
        value.rating_tree_match_method as RatingTreeMatchMethod
      )) &&
    (value.rating_tree_match_evidence === undefined ||
      value.rating_tree_match_evidence === null ||
      (typeof value.rating_tree_match_evidence === "string" &&
        value.rating_tree_match_evidence.trim().length > 0)) &&
    (value.standard_defect_indicator_id === undefined ||
      value.standard_defect_indicator_id === null ||
      (typeof value.standard_defect_indicator_id === "string" &&
        value.standard_defect_indicator_id.trim().length > 0)) &&
    photoReferences !== null &&
    photoReferences.every(isValidPhotoReference) &&
    new Set(photoReferences.map((item) => (item as PhotoReference).photo_number)).size ===
      photoReferences.length
  );
}

function isValidPhotoCandidate(value: unknown): boolean {
  return (
    isRecord(value) &&
    hasValidConfidence(value) &&
    hasRequiredArrayMembers(value, ["warnings"]) &&
    hasRequiredObjectMembers(value, ["extracted_file", "source_ref"]) &&
    isValidSourceRef(value.source_ref)
  );
}

function isValidComparisonCandidate(value: unknown): boolean {
  if (!isRecord(value) || !hasValidConfidence(value) || getRequiredArray(value, "warnings") === null) {
    return false;
  }
  if (value.match_basis === undefined || value.match_basis === null) {
    return true;
  }
  return isRecord(value.match_basis) && isNumberFromZeroToOne(value.match_basis.location_similarity);
}

export function isBridgeAnnualInspectionData(value: unknown): value is BridgeAnnualInspectionData {
  if (!isRecord(value)) {
    return false;
  }

  // 契约名和版本先过关，避免旧格式数据进入新校对流程。
  const contract = getRequiredObject(value, "contract");
  if (contract === null) {
    return false;
  }
  if (contract.name !== "BridgeAnnualInspectionData" || contract.version !== "3.0") {
    return false;
  }

  if (hasOwn(value, "ratings")) {
    return false;
  }

  const importContext = getRequiredObject(value, "import_context");
  const bridgeCheck = getRequiredObject(value, "bridge_check");
  const inspection = getRequiredObject(value, "inspection");
  if (importContext === null || bridgeCheck === null || inspection === null) {
    return false;
  }

  const defects = getRequiredArray(value, "defects");
  const photos = getRequiredArray(value, "photos");
  const comparisonCandidates = getRequiredArray(value, "comparison_candidates");
  const reportTextCandidates = getRequiredArray(value, "report_text_candidates");
  const warnings = getRequiredArray(value, "warnings");
  const errors = getRequiredArray(value, "errors");
  if (
    defects === null ||
    photos === null ||
    comparisonCandidates === null ||
    reportTextCandidates === null ||
    warnings === null ||
    errors === null
  ) {
    return false;
  }

  if (getRequiredArray(bridgeCheck, "warnings") === null) {
    return false;
  }

  return (
    defects.every(isValidDefectCandidate) &&
    photos.every(isValidPhotoCandidate) &&
    comparisonCandidates.every(isValidComparisonCandidate)
  );
}
